### DISCUSSION TITLE
Run on ARM Architecture

### DISCUSSION THREAD

**Discussion by briis** [briis/hass-weatherflow2mqtt#10] [Announcements]
Currently the Prebuild Container will only work if you run Docker on the Intel Architecture, but if you are running on something different, then @simonransom wrote a very detailed set of instructions on how to get this to run on a different Architecture than Intel. I have copied the instructions Simon made from the Github Issue below:

Thanks a lot Simon. 👍🏻

Here's how I got it working...

To create PI Image, steps 1-3 as per @briis' instructions:

On a Pi with Docker installed, download the source from Github to the Pi.
Unpack the ZipFile, and go to the directory where you find the file Dockerfile
Build the container locally by issuing this command: docker build -t weatherflow2mqtt . (don’t forget the . at the end of the command).
I now wanted to get this ready to use in Portainer, but couldn’t work our how to move files from one system to another, so I created an account at https://hub.docker.com/ and uploaded it there as follows:

On the Pi used above ran command: docker login
Tagged PI image with version number with command: docker tag weatherflow2mqtt simonranson/weatherflow2mqtt:v1
Pushed PI image with command: docker push simonranson/weatherflow2mqtt:v1
I’ll delete this image once from docker.com I find a way to move files in Portainer.

Now onto the HA machines. This assumes the Mosquitto broker is already installed in HA

In HA, install the “Portainer” and “Samba Share” add-ons.
In the Samba Share configuration, set a username and password, then start the add-on
On another machine on the network, browse to the IP/hostname of your HA server, and enter the credentials as per step 2.
Create a folder for the integration to use. I created a folder called weatherflow under \config\custom_components.
Start the Portainer add-on (you do need to disable protection mode to do this).
You can now open up the Portainer GUI and prepare the volume

In Portainer, navigate to “Volumes” in the left hand column and click on the blue “+ Add volume”
Give it a name, I used Weatherflow
Turn on CIFS
In the Address box, enter the IP/Hostname of your HA server
In the Share box, the path from step 4 above (watch the “/” direction here), for my setup this was /config/custom_components/weatherflow
Username and Password are as per step 2 above.
Now to add the container…

Navigate to “Containers” on the left, and click the blue “+ Add container”
Give it a name such as weatherflow2mqtt.
Make sure “DockerHub” is selected in the Registry box, then use simonranson/weatherflow2mqtt:v1 as the image name
For now, just accept all the defaults for everything else and click “Deploy the container”
After a little while, the container will install, you will be able to see this by clicking the Containers link on the left. It’ll now need configuring.

Navigate to “Containers” on the left, and click theweatherflow2mqtt container.
Click the blue “Duplicate/Edit” button in the top right.
Scroll down to the “Advanced Container Settings”
Click “Volume” and click “map additional volume”
Make sure “Volume” and “Writable” are selected on the right (darker blue colour)
In Container enter: /usr/local/config
In Volume use the drop-down to select your volume created in step 7
Click “Network” and use the drop-down to select “host”
Click “Restart policy” and select “Unless Stopped”
Finally, head back to “Env” and set all your variables as per the Github instructions, ignore anything above TZ.
Last thing to do now is press the “Deploy the container” button. Press “Replace” to the are you sure message.
Now if you look back in the containers section, you’ll see the state green/running. If you click the left most quick-action link you should see some activity. You can run through steps 16+ to go back in and set the debugs to true if you are having issues – don’t forget step 26 to redeploy if you make any changes.

Hope this helps anybody wanting a play.

**Comment by simonranson** [comment]
For now I am keeping the latest ARM builds at https://hub.docker.com/repository/docker/simonranson/weatherflow2mqtt. Deploy the latest version.

Cheers
Simon

**Comment by muxel66** [comment]
I tried to set it up as described on my Home Assistant OS Odroid. After getting root access to the machine the setup went as specified. However after starting the container I get the following error in the logs (repeating steadily):


> INFO:__main__:SETTING UP UV Level SENSOR
> 
> INFO:__main__:SETTING UP Beaufort Scale SENSOR
> 
> INFO:__main__:SETTING UP Weather SENSOR
> 
> INFO:sqlite:Upgrading the database to version 1
> 
> ERROR:sqlite:Could not create SQL Table. Error: database is locked
> 
> ERROR:sqlite:Could not Insert data in table high_low. Error: no such table: high_low
> 
> INFO:sqlite:Upgrading the database to version 2...
> 
> ERROR:sqlite:An undefined error occured. Error message: no such table: high_low
> 
> ERROR:sqlite:Could not access storage data. Error: no such table: storage
> 
> ERROR:sqlite:Could not access high_low data. Error: no such table: high_low

I assume the problem is the locked db. I did a chmod 666 to the db, but it did not help. 
I would appreciate any help, especially as I am not very fluent in linux and the command line...

Thanks
Axel

**Reply by briis** [reply]
@simonranson Did you do anything specific to ensure that you could write to the weatherflow2mqtt.db?

**Reply by muxel66** [reply]
Only the mentioned CHMOD. Is that not enough? Sorry, new with all of that.

**Reply by briis** [reply]
Could you try and do a

ls -l 

in the directory and post the output here

**Reply by muxel66** [reply]
Sure:

# ls -l
-rwxrwxrwx    1 root     root             0 Jul  7 11:04 weatherflow2mqtt.db

**Reply by briis** [reply]
Try and delete the database. It is somehow corrupted.

`sudo rm weatherflow2mqtt.db`

Stop the container before you do this. And then start again after the deletion.

**Reply by muxel66** [reply]
I tried that several times with no success. Also started from scratch with everything with no improvement. I guess something does not agree with the HA OS system...

**Reply by briis** [reply]
I am tight up with work right now, but hopefully I will have a change to look at this later this week.

**Comment by ghgeiger** [comment]
This post is just a warning to anyone trying to follow the directions to get this running with the portainer hass add-on v 2.0.0, that it probably will not work.

(Excuse my ignorance ahead of time, this is my first time dealing with docker or portainer directly.)

I've been working for days to get this working, but I've run into a roadblock where portainer won't allow me to add a new container, it appears to have to do with this [issue](https://community.home-assistant.io/t/portainer-v2-6-2-unable-to-start-containers/332356/32). It seems as though that the latest portainer add-on for home assistant (2.0.0, portainer v. 2.6.2) has a bug that wont let any containers be added or existing ones started. They've listed a workaround, basically someone cloned the older version, and I'm about to delve into trying to get that to work.

One thing I would really like to know, though, is if anyone is working to make hass_weatherflow2mqqt into an actual add-on that can be installed directly in supervisor? I don't know what all goes into it, but I would think it wouldn't be too difficult since the docker already exists, and it would've saved me many hours and many more lost hairs.

Thanks
Greg

**Reply by briis** [reply]
Just tried this also on a HA Blue, and Portainer could not create the Container. If the workaround works, please post it here.

If I am not working on making this a Hass Add-On, I don't think anyone is - and I am not. Not saying it will never happen, but I don't have the time right now, to learn how to make an official Add-On.

Any volunteers - just ping me.

**Reply by ghgeiger** [reply]
The workaround is a backup of the previous Portainer that someone cloned. The clone is located here:

[https://github.com/MaxWinterstein/homeassistant-addons](https://github.com/MaxWinterstein/homeassistant-addons)

However (and I may totally be doing something wrong), I can not get supervisor to accept the directory above, or the one that contains just portainer, as a valid add-on repository. So I am at yet another roadblock. 

I was hoping to be able to have everything HA running on a single pi at my house, I guess I can either live with the older UDP HA integration, or set up another pi just to run this in docker. 

Thanks for all of your hard work on this, and your previous work on SmartweatherUDP.
